up-discovery-android/service/src/main/java/org/eclipse/uprotocol/core/udiscovery/Notifier.java

```diff
@@ -0,0 +1,163 @@
+        });
+    }
+
+    public void notifyObserversAddNodes(List<UUri> nodePath, List<UUri> addedNodes) {
+        if (nodePath.isEmpty()) {
+            Log.w(TAG, join(Key.MESSAGE, "notifyObserversAddNodes nodePath is Empty"));
+            return;
+        }
+        final List<UUri> observerList = getListOfObservers(nodePath);
+        final UUri parent = nodePath.get(nodePath.size() - 1);
+        for (UUri observer : observerList) {
+            for (UUri child : addedNodes) {
+                if (DEBUG) {
+                    Log.d(TAG, join(Key.MESSAGE, "notify",
+                            Key.URI, toLongUri(child),
+                            PARENT_URI, toLongUri(parent),
+                            OBSERVER_URI, toLongUri(observer)));
+                }
+                buildNotifications(child, parent, Operation.ADD, observer);
+            }
+        }
+    }
+
+    private List<UUri> getListOfObservers(List<UUri> nodePath) {
+        ArrayList<UUri> observerList = new ArrayList<>();
+        for (UUri nodePathUri : nodePath) {
+            for (Map.Entry<UUri, Set<UUri>> entry : mObserverManager.getObserverMap().entrySet()) {
+                if (nodePathUri.equals(entry.getKey())) {
+                    observerList.addAll(entry.getValue());
+                }
+            }
+        }
+        return observerList;
+    }
+
+    private void buildNotifications(UUri nodeUri, UUri parentUri, Operation operation,
+                                    UUri observerUri) {
+        final String child = toLongUri(nodeUri);
+        final String parent = toLongUri(parentUri);
+        final Notification notification = Notification.newBuilder()
+                .setOperation(operation)
+                .setUri(child)
+                .setParentUri(parent)
+                .build();
+        sendNotification(observerUri, notification);
+    }
+
+    /**
+     * sendNotification - Notify the observer of the change to a node by sending
+     * Notification on topic: core.udiscovery/3/nodes#Notification
+     *
+     * @param observer     - observer to whom the notification should be sent to
+     * @param notification - Notification containing the details of the node and operation
+     */
+    private void sendNotification(@NonNull UUri observer, @NonNull Notification notification) {
+        Log.d(TAG, join(OBSERVER_URI, observer.toString(), NOTIFICATION,
+                notification));
+        // TODO
+//        final Any payload = Any.pack(notification);
+//        final String sink = toLongUri(observer);
+//        final CloudEvent notificationEvent =
+//                CloudEventFactory.notification(TOPIC_NODE_NOTIFICATION, sink, payload,
+//                        UCloudEventAttributes.empty());
+//        if (DEBUG) {
+//            Log.d(LOG_TAG, Key.EVENT, "sendNotification", Key.MESSAGE, stringify(notificationEvent));
+//        }
+//        final Status status = mUltifiLink.publish(notificationEvent);
```

Comment:
Please remove ultifi references and if this code is not used, remove it

Reply:
Remove comment file.